File: src/forager/library/playtime.py
"""Per-game playtime tracking.

Sessions are recorded for games launched through forager. A session keeps
accumulating while the spawned child process is still alive; ``last_played``
is stamped at launch time, so the Recently Played row works even for Steam
games (the ``steam://rungameid`` command exits immediately, so Steam playtime
is not accumulated — only the last-played stamp). State persists to
``playtime.json`` next to ``settings.json``.
"""
from __future__ import annotations
import json
import subprocess
import time
from pathlib import Path

from forager.core.game import Game
from forager.core.paths import playtime_file


def game_key(game: Game) -> str:
    """Stable identity for a game across scans.

    Steam games key on the app id; everything else keys on the resolved path
    (paths are unique per source, and standalone/Minecraft games have no id).
    """
    if game.app_id:
        return f"steam:{game.app_id}"
    try:
        path = str(game.path.resolve())
    except (OSError, AttributeError):
        path = str(game.path) if game.path is not None else game.name
    return f"{game.source.name.lower()}:{path}"


def format_playtime(seconds: float) -> str:
    if seconds < 60:
        return f"{int(seconds)} s"
    if seconds < 3600:
        return f"{int(seconds // 60)} min"
    return f"{seconds / 3600:.1f} h"


class PlaytimeStore:
    def __init__(self, path: Path | None = None):
        self.path = Path(path) if path is not None else playtime_file()
        self._data: dict[str, dict] = {}
        self.load()

    def load(self) -> None:
        data: dict = {}
        try:
            if self.path.is_file():
                raw = json.loads(self.path.read_text("utf-8"))
                if isinstance(raw, dict):
                    data = raw
        except (OSError, json.JSONDecodeError):
            data = {}
        self._data = data

    def save(self) -> None:
        try:
            self.path.parent.mkdir(parents=True, exist_ok=True)
            self.path.write_text(json.dumps(self._data, indent=2) + "\n", "utf-8")
        except OSError:
            pass

    def get(self, key: str) -> dict:
        entry = self._data.get(key)
        if entry is None or not isinstance(entry, dict):
            entry = {}
        return entry

    def playtime(self, key: str) -> float:
        return float(self.get(key).get("playtime", 0) or 0)

    def last_played(self, key: str) -> float:
        return float(self.get(key).get("last_played", 0) or 0)

    def touch(self, key: str, when: float | None = None) -> None:
        """Stamp a launch without changing accumulated playtime."""
        entry = dict(self.get(key))
        entry["last_played"] = time.time() if when is None else when
        entry["playtime"] = float(entry.get("playtime", 0) or 0)
        self._data[key] = entry

    def add(self, key: str, seconds: float) -> None:
        if seconds <= 0:
            return
        entry = dict(self.get(key))
        entry["playtime"] = float(entry.get("playtime", 0) or 0) + seconds
        entry["last_played"] = float(entry.get("last_played", 0) or 0)
        self._data[key] = entry


class PlaytimeTracker:
    """Tracks live sessions and exposes recently-played ranking.

    A session is created on launch and accumulates time on each ``tick``
    while the spawned process is alive; when it exits the final chunk is
    added and the session is dropped. ``flush`` ends every session and saves.
    """

    def __init__(self, store: PlaytimeStore | None = None):
        self._store = store if store is not None else PlaytimeStore()
        self._sessions: dict[str, dict] = {}

    @property
    def store(self) -> PlaytimeStore:
        return self._store

    def has_sessions(self) -> bool:
        return bool(self._sessions)

    def begin(self, game: Game, proc: subprocess.Popen | None) -> None:
        """Start a session. ``proc`` may be None when the game can't be
        tracked (Steam: the CLI command returns immediately)."""
        key = game_key(game)
        self._store.touch(key)
        if proc is not None:
            self._sessions[key] = {"proc": proc, "last": time.time()}
        self._store.save()

    def tick(self) -> bool:
        """Accumulate elapsed time for live sessions. Returns True when the
        store changed (so the UI can refresh)."""
        now = time.time()
        dirty = False
        for key, sess in list(self._sessions.items()):
            proc = sess["proc"]
            running = proc is not None and proc.poll() is None
            if running:
                self._store.add(key, max(0.0, now - sess["last"]))
                sess["last"] = now
                dirty = True
            else:
                if proc is not None:
                    self._store.add(key, max(0.0, now - sess["last"]))
                    dirty = True
                del self._sessions[key]
        if dirty:
            self._store.save()
        return dirty

    def flush(self) -> None:
        self.tick()
        self._sessions.clear()
        self._store.save()

    def is_running(self, game: Game) -> bool:
        key = game_key(game)
        sess = self._sessions.get(key)
        if sess is None:
            return False
        proc = sess["proc"]
        return proc is not None and proc.poll() is None

    def stop(self, game: Game) -> bool:
        """Terminate the running session for ``game``, if any.

        Returns True when a live process was actually terminated. The final
        time chunk is recorded and the session is dropped.
        """
        key = game_key(game)
        sess = self._sessions.get(key)
        if sess is None:
            return False
        proc = sess["proc"]
        if proc is not None and proc.poll() is None:
            proc.terminate()
            try:
                proc.wait(timeout=5)
            except subprocess.TimeoutExpired:
                proc.kill()
            self._store.add(key, max(0.0, time.time() - sess["last"]))
        del self._sessions[key]
        self._store.save()
        return True

    def recently_played(self, games: list[Game], limit: int = 8) -> list[Game]:
        """Games with a last-played stamp, newest first."""
        ranked = [
            g for g in games
            if self._store.last_played(game_key(g)) > 0
        ]
        ranked.sort(key=lambda g: self._store.last_played(game_key(g)), reverse=True)
        return ranked[:limit]

File: src/library/Playtime.cpp
#include "library/Playtime.h"
#include "utils/Json.h"
#include "core/Paths.h"

#include <QJsonObject>
#include <QElapsedTimer>

static float nowSeconds()
{
    static QElapsedTimer timer;
    if (!timer.isValid()) timer.start();
    return timer.elapsed() / 1000.0f;
}

// --- PlaytimeStore ---

PlaytimeStore::PlaytimeStore(const QString& path, QObject* parent)
    : QObject(parent)
    , m_path(path.isEmpty() ? paths::playtimeFile() : path)
{
    load();
}

void PlaytimeStore::load()
{
    auto data = json::readObject(m_path);
    if (!data) return;

    QMutexLocker lock(&m_mutex);
    m_data = *data;
}

void PlaytimeStore::save()
{
    QJsonObject snapshot;
    {
        QMutexLocker lock(&m_mutex);
        snapshot = m_data;
    }

    json::writeObject(m_path, snapshot);
}

float PlaytimeStore::playtime(const QString& key) const
{
    QMutexLocker lock(&m_mutex);
    return m_data[key].toObject()["playtime"].toDouble(0.0);
}

float PlaytimeStore::lastPlayed(const QString& key) const
{
    QMutexLocker lock(&m_mutex);
    return m_data[key].toObject()["last_played"].toDouble(0.0);
}

void PlaytimeStore::mutateEntry(const QString& key, const std::function<void(QJsonObject&)>& fn)
{
    QMutexLocker lock(&m_mutex);
    QJsonObject entry = m_data[key].toObject();
    fn(entry);
    m_data[key] = entry;
}

void PlaytimeStore::touch(const QString& key)
{
    mutateEntry(key, [](QJsonObject& entry) {
        entry["last_played"] = nowSeconds();
        if (!entry.contains("playtime")) entry["playtime"] = 0.0;
    });
}

void PlaytimeStore::add(const QString& key, float seconds)
{
    if (seconds <= 0) return;
    mutateEntry(key, [seconds](QJsonObject& entry) {
        entry["playtime"] = entry["playtime"].toDouble(0.0) + seconds;
        if (!entry.contains("last_played")) entry["last_played"] = 0.0;
    });
}

QString PlaytimeStore::gameKey(const Game& game)
{
    if (!game.appId().isEmpty())
        return "steam:" + game.appId();

    QString path = game.path();
    if (!path.isEmpty()) {
        // Canonical path would be better, but may not exist at lookup time
        return game.sourceName().toLower() + ":" + path;
    }
    return game.sourceName().toLower() + ":" + game.name();
}

QString PlaytimeStore::formatPlaytime(float seconds)
{
    if (seconds < 60)
        return QStringLiteral("%1 s").arg(static_cast<int>(seconds));
    if (seconds < 3600)
        return QStringLiteral("%1 min").arg(static_cast<int>(seconds / 60));
    return QStringLiteral("%1 h").arg(seconds / 3600.0f, 0, 'f', 1);
}

// --- PlaytimeTracker ---

PlaytimeTracker::PlaytimeTracker(PlaytimeStore* store, QObject* parent)
    : QObject(parent)
    , m_store(store)
{
}

bool PlaytimeTracker::hasSessions() const
{
    return !m_sessions.isEmpty();
}

void PlaytimeTracker::begin(const Game& game, QProcess* proc)
{
    QString key = PlaytimeStore::gameKey(game);
    m_store->touch(key);
    if (proc) {
        m_sessions[key] = {proc, nowSeconds()};
    }
    m_store->save();
}

bool PlaytimeTracker::tick()
{
    float current = nowSeconds();
    bool dirty = false;

    auto it = m_sessions.begin();
    while (it != m_sessions.end()) {
        auto& sess = it.value();
        if (flushSession(it.key(), sess, current)) dirty = true;

        if (sess.proc && sess.proc->state() != QProcess::NotRunning) {
            sess.last = current;
            ++it;
        } else {
            it = m_sessions.erase(it);
        }
    }

    if (dirty) m_store->save();
    return dirty;
}

bool PlaytimeTracker::flushSession(const QString& key, Session& sess, float now)
{
    float elapsed = now - sess.last;
    if (elapsed <= 0) return false;
    m_store->add(key, elapsed);
    return true;
}

void PlaytimeTracker::flush()
{
    tick();
    m_sessions.clear();
    m_store->save();
}

bool PlaytimeTracker::isRunning(const Game& game) const
{
    QString key = PlaytimeStore::gameKey(game);
    auto it = m_sessions.find(key);
    if (it == m_sessions.end()) return false;
    return it->proc && it->proc->state() != QProcess::NotRunning;
}

bool PlaytimeTracker::stop(const Game& game)
{
    QString key = PlaytimeStore::gameKey(game);
    auto it = m_sessions.find(key);
    if (it == m_sessions.end()) return false;

    auto& sess = it.value();
    if (sess.proc && sess.proc->state() != QProcess::NotRunning) {
        flushSession(key, sess, nowSeconds());

        sess.proc->terminate();
        if (!sess.proc->waitForFinished(2000)) {
            sess.proc->kill();
            sess.proc->waitForFinished(1000);
        }
    }

    m_sessions.erase(it);
    m_store->save();
    return true;
}

QList<Game> PlaytimeTracker::recentlyPlayed(const QList<Game>& games, int limit) const
{
    QList<Game> ranked;
    for (const auto& g : games) {
        QString key = PlaytimeStore::gameKey(g);
        if (m_store->lastPlayed(key) > 0) {
            ranked.append(g);
        }
    }

    std::sort(ranked.begin(), ranked.end(), [this](const Game& a, const Game& b) {
        return m_store->lastPlayed(PlaytimeStore::gameKey(a))
             > m_store->lastPlayed(PlaytimeStore::gameKey(b));
    });

    while (ranked.size() > limit) ranked.removeLast();
    return ranked;
}
